File: Week_04/walking_robot_simulation/Solution.java
package walking_robot_simulation;

import java.util.HashSet;
import java.util.Set;

class Solution {
    public int robotSim(int[] commands, int[][] obstacles) {
        Set<String> set = new HashSet<>();
        for (int[] obs : obstacles) {
            set.add(obs[0] + "," + obs[1]);
        }
        int[][] dirs = new int[][]{{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        int dir = 0, x = 0, y = 0, result = 0;
        for (int c : commands) {
            if (c == -1) {
                dir = dir == 3 ? 0 : dir + 1;
            } else if (c == -2) {
                dir = dir == 0 ? 3 : dir - 1;
            } else {
                int[] xy = dirs[dir];
                while (c-- > 0 && !set.contains((x + xy[0]) + "," + (y + xy[1]))) {
                    x += xy[0];
                    y += xy[1];
                }
            }
            result = Math.max(result, x * x + y * y);
        }
        return result;
    }
}

File: Week_04/walking_robot_simulation/Solution.py
class Solution:
    def robotSim(self, commands: List[int], obstacles: List[List[int]]) -> int:

        obs = set(map(tuple, obstacles))
        dir = ((0,1), (1,0), (0,-1), (-1,0))

        d, x, y, max_dis = 0, 0, 0, 0

        for step in commands:
            if step == -1:
                d = d + 1 if d < 3 else 0
            elif step == -2:
                d = d - 1 if d > 0 else 3
            else:
                for _ in range(step):
                    if (x + dir[d][0], y + dir[d][1]) not in obs:
                        x, y = x + dir[d][0], y + dir[d][1]
                max_dis = max(max_dis, x*x + y*y)
        return max_dis
